cpp/src/arrow/util/atomic_util.h

```diff
@@ -0,0 +1,157 @@
+            return _InterlockedExchangeAdd(addr, val);
+        }
+
+#if _WIN64
+        template <typename T, typename std::enable_if<sizeof(T) == 8>::type>
+        inline T AtomicFetchAdd(T *addr, T &val, std::memory_order /*order*/) noexcept
+        {
+            static_assert(std::is_integral<T>::value, "AtomicFetchAdd can only be used on integral types");
+            return _InterlockedExchangeAdd64(addr, val);
+        }
+#else
+        template <typename T, typename std::enable_if<sizeof(T) == 8>::type>
+        inline T AtomicFetchAdd(T *addr, T &val, std::memory_order /*order*/) noexcept
+        {
+            static_assert(std::is_integral<T>::value, "AtomicFetchAdd can only be used on integral types");
+            _ReadWriteBarrier();
+            T expected = *addr;
+            for(;;)
+            {
+                T new_val = expected + val;
+                T prev = _InterlockedCompareExchange64(addr, new_val, expected);
+                if(prev == expected)
+                    return prev;
+                expected = prev;
+            }
+        }
+#endif
+
+        template <typename T, typename std::enable_if<sizeof(T) == 1>::type>
+        inline T AtomicFetchSub(T *addr, T &val, std::memory_order /*order*/) noexcept
+        {
+            static_assert(std::is_integral<T>::value, "AtomicFetchSub can only be used on integral types");
+            return _InterlockedExchangeAdd8(addr, -val);
+        }
+
+        template <typename T, typename std::enable_if<sizeof(T) == 2>::type>
+        inline T AtomicFetchSub(T *addr, T &val, std::memory_order /*order*/) noexcept
+        {
+            static_assert(std::is_integral<T>::value, "AtomicFetchSub can only be used on integral types");
+            return _InterlockedExchangeAdd16(addr, -val);
+        }
+
+        template <typename T, typename std::enable_if<sizeof(T) == 4>::type>
+        inline T AtomicFetchSub(T *addr, T &val, std::memory_order /*order*/) noexcept
+        {
+            static_assert(std::is_integral<T>::value, "AtomicFetchSub can only be used on integral types");
+            return _InterlockedExchangeAdd(addr, -val);
+        }
+
+#if _WIN64
+        template <typename T, typename std::enable_if<sizeof(T) == 8>::type>
+        inline T AtomicFetchSub(T *addr, T &val, std::memory_order /*order*/) noexcept
+        {
+            static_assert(std::is_integral<T>::value, "AtomicFetchSub can only be used on integral types");
+            return _InterlockedExchangeAdd64(addr, -val);
+        }
+#else
+        template <typename T, typename std::enable_if<sizeof(T) == 8>::type>
+        inline T AtomicFetchSub(T *addr, T &val, std::memory_order /*order*/) noexcept
+        {
+            static_assert(std::is_integral<T>::value, "AtomicFetchSub can only be used on integral types");
+            return AtomicFetchAdd(addr, -val, std::memory_order_seq_cst);
+        }        
+#endif
+#endif
```

Comment:
Why do you need these instead of `std::atomic`?

Reply:
Sometimes you don't want to always have an atomic, and only do an atomic operation in a specific situation. In other words, using these you can upgrade a non-atomic to an atomic temporarily, but you can't downgrade an atomic to a non-atomic. C++20 has `std::atomic_ref` to solve this, but we're still on 11.